packages/react/src/utils/floating-ui/useClick.ts

```diff
@@ -0,0 +1,122 @@
+'use client';
+import * as React from 'react';
+import type { ElementProps, FloatingRootContext } from '@floating-ui/react';
+import { isMouseLikePointerType } from '@floating-ui/react/utils';
+
+export interface UseClickProps {
+  /**
+   * Whether the Hook is enabled, including all internal Effects and event
+   * handlers.
+   * @default true
+   */
+  enabled?: boolean;
+  /**
+   * The type of event to use to determine a “click” with mouse input.
+   * Keyboard clicks work as normal.
+   * @default 'click'
+   */
+  event?: 'click' | 'mousedown';
+  /**
+   * Whether to toggle the open state with repeated clicks.
+   * @default true
+   */
+  toggle?: boolean;
+  /**
+   * Whether to ignore the logic for mouse input (for example, if `useHover()`
+   * is also being used).
+   * @default false
+   */
+  ignoreMouse?: boolean;
+  /**
+   * If already open from another event such as the `useHover()` Hook,
+   * determines whether to keep the floating element open when clicking the
+   * reference element for the first time.
+   * @default true
+   */
+  stickIfOpen?: boolean;
+}
+
+/**
+ * Opens or closes the floating element when clicking the reference element.
+ * @see https://floating-ui.com/docs/useClick
+ */
+export function useClick(context: FloatingRootContext, props: UseClickProps = {}): ElementProps {
+  const { open, onOpenChange, dataRef } = context;
+  const {
+    enabled = true,
+    event: eventOption = 'click',
+    toggle = true,
+    ignoreMouse = false,
+    stickIfOpen = true,
+  } = props;
+
+  const pointerTypeRef = React.useRef<'mouse' | 'pen' | 'touch'>(undefined);
+  const frameRef = React.useRef(-1);
+
+  React.useEffect(() => {
+    return () => {
+      if (frameRef.current !== -1) {
+        cancelAnimationFrame(frameRef.current);
+        frameRef.current = -1;
+      }
+    };
+  }, []);
+
+  const reference: ElementProps['reference'] = React.useMemo(
+    () => ({
+      onPointerDown(event) {
+        pointerTypeRef.current = event.pointerType;
+      },
+      onMouseDown(event) {
+        const pointerType = pointerTypeRef.current;
+
+        // Ignore all buttons except for the "main" button.
+        // https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
+        if (
+          event.button !== 0 ||
+          eventOption === 'click' ||
+          (isMouseLikePointerType(pointerType, true) && ignoreMouse)
+        ) {
+          return;
+        }
+
+        const nextOpen = !(
+          open &&
+          toggle &&
+          (dataRef.current.openEvent && stickIfOpen
+            ? dataRef.current.openEvent.type === 'mousedown'
+            : true)
+        );
+        // Wait until focus is set on the element. This is an alternative to
+        // `event.preventDefault()` to avoid :focus-visible from appearing when using a pointer.
+        frameRef.current = requestAnimationFrame(() => {
+          onOpenChange(nextOpen, event.nativeEvent, 'click');
+        });
```

Comment:
I don't like adding async behavior as it makes testing harder, but this fixes the issue